### ISSUE TITLE
very old (1y+) notifications are shown

### ISSUE THREAD

**Issue by nehemiagurl** [tuskyapp/Tusky#3689]
Starting today, very old notifications in two of my instances are shown.  Clicking the bell icon to auto-scroll all the way to the top doesn't help: it just loads newer, but still 1y+ old notifications. Same with the three dot refresh button.
Closing the app and force stopping did not assist. Switching accounts back and forth didn't either.
![Screenshot_20230525-181258_Tusky](https://github.com/tuskyapp/Tusky/assets/109424320/3c1ed0cc-b37c-44c5-bdf2-25614ab48691)
![Screenshot_20230525-181340_Tusky](https://github.com/tuskyapp/Tusky/assets/109424320/4f576203-9d44-4c61-be3b-40f1c616cc61)

~Weirdly enough this doesn't happen in all of my accounts, just two of them for unclear reasons. It's also not the admin privileges that give the sign-up and report notifications that are to blame, because I am also sign in with the main admin account for both instances and in them this doesn't happen.~
EDIT: okay now that I switched away from them and back it does happen in them as well. so it seems the admin-type notifications (signups and reports) are to blame.

* * * *
- Tusky Version: 22.0 beta 5
- Android Version: 12
- Android Device: Galaxy A12
- Mastodon instance (if applicable): tooot.im, toki.social

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by nikclayton** [comment]
Thanks for the report, I've been trying to track this down. And sorry for the hassle.

You mentioned you're signed in to multiple accounts.

Do you remember if the following has happened:

1. You were using one of your accounts.
2. You received an **Android** notification about a message in one of the other accounts
3. From the **Android** notification there was a "Compose" option, so you could reply directly from the notification.
4. Did you use that "Compose" feature

?

If you did I think this is a bug I've identified and am fixing in https://github.com/tuskyapp/Tusky/pull/3688.

If you didn't then something else is going on.

**Comment by nikclayton** [comment]
# tl;dr

The only thing I can think of is that there's another bug like https://github.com/tuskyapp/Tusky/pull/3688 where the active account is being changed. Something like the following.

- User is on two servers, A and B. A is extremely busy, thousands of users, lots of activity. B is quiet.
- **If** notification IDs on a server are implemented as monotonically increasing ints, and are per-server, not per user, this would mean that the notification IDs on server are A very high, and the notification IDs on server B are very low(*)
- The user is using their account on server B, where the notification IDs are low.
- They do something that changes `accountManager.activeAccount` to point to their account on server A, **and** `NotificationsViewModel` is not restarted.
- The user navigates away from the Notifications tab. Their very low `lastNotificationId` for server B is saved to the record for server A.
- They switch to their server A account.
- They see notifications from years ago, because the notification ID is too old.

(*) "very low" here means "Interpreted by the Mastodon server as being very low". See the GoToSocial example at the bottom of this comment for an example where the ID string is very long but is interpreted as a low number.

But I might be wrong...

# Observed symptoms

User opens the Notifications tab, and their oldest notifications are loaded instead of the most recent / the notifications around their most recent reading position.

# Theories

Either the app is:

1. requesting the ID of a very old notification.

2. making a request that the server doesn't understand, and the server is responding with a page of very old notifications.

3. initially loading the correct notification, but loading all the older ones in the background, and scrolling to the bottom

I think option 3 can be ruled out. If that was happening then there would be a delay, as loading pages and pages of old notifications from the server would take some time (probably many seconds) before any scrolling could happen, and that's not being reported by users.

Also, Tusky would have loaded all the intermediate notifications, so tapping the notification tab would work to take you to the most recent page. Again, that's not what people are reporting.

So it's option 1 or option 2.

# Option 1 investigation

Is Tusky requesting the ID of a very old notification?

The initialKey for loading notifications comes from `accountManager.activeAccount?.lastNotificationId`, which is a `String?`.

If that's `null` then `NotificationsPagingSource.getInitialPage()` returns the most recent notifications, by making the following API request (`exclude_types` will vary depending on the user's notification filters).

```
https://mastodon.social/api/v1/notifications?limit=30&exclude_types%5B%5D=follow_request
```

There are no `min_id` or `max_id` parameters.

If `lastNotificationId` exists then that is used instead.

The `lastNotificationId` property is only written to in one place in the code, `NotificationsViewModel` where the `SaveVisibleId` action is handled. And that action is only sent by `NotificationsFragment` when the user stops scrolling or leaves the tab.

As far as I can tell this is always saving the value of a valid Notification ID.

Worth noting -- `lastNotificationId` is **not** affected by the notification marker that might be set using the Mastodon API. That's used for Android Notifications only, not the Notifications Tab.

`lastNotificationId` is saved as part of the active account.

# Option 2 investigation

Is Tusky making an invalid request, and getting a page of oldest notifications?

I did some tests with possible invalid values for `lastNotificationId`.

- `""` (empty string). This throws an exception, because there's no value for the ID so the API returns a page of the most recent notifications and the code is expecting a single notification. That's fine.

- `"1"`. As expected, this returns the very oldest notification.

- `"-1"`. This also returns the very oldest notifications.

- `"a"`. This returns the newest notifications. Which is odd, since it sorts well before the IDs of the most recent notifications (which have IDs around 152563233).

- `"01G65Z755AFWAKHE12NY0CQ9FH"` (a GoToSocial-like ID). That returns the oldest notifications. Which is also odd, given the behaviour with `a`.

So if `lastNotificationId` is getting corrupted in some way then this could fit the observed behaviour.

**Comment by nehemiagurl** [comment]
One thing I don't understand about these theories: why does this only happen in accounts with admin/moderator privileges? Even when comparing two accounts on the same instance - the one with the admin privileges will have this bug, the one without them won't.

**Comment by nikclayton** [comment]
> One thing I don't understand about these theories: why does this only happen in accounts with admin/moderator privileges?

I don't know that it does yet. You're the first person to mention that difference between your accounts in the very small number of reports we've received so far.

That's not to say it might not be an important detail, but right now I just don't have enough information.

**Comment by nikclayton** [comment]
The investigation in https://github.com/tuskyapp/Tusky/issues/3696 put me on the right track, I think.

Apply this diff to get some extra logging:

```diff
diff --git a/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsFragment.kt b/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsFragment.kt
index ab5c1b53b..0b55ac47a 100644
--- a/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsFragment.kt
+++ b/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsFragment.kt
@@ -119,6 +119,7 @@ class NotificationsFragment :
         container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View {
+        Log.d(TAG, "onCreateView, active account = ${viewModel.activeAccount?.id}")
         return inflater.inflate(R.layout.fragment_timeline_notifications, container, false)
     }

@@ -465,6 +466,7 @@ class NotificationsFragment :

     override fun onPause() {
         super.onPause()
+        Log.d(TAG, "onPause, active account = ${viewModel.activeAccount?.id}")

         // Save the ID of the first notification visible in the list
         val position = layoutManager.findFirstVisibleItemPosition()
@@ -622,7 +624,7 @@ class NotificationsFragment :
     }

     companion object {
-        private const val TAG = "NotificationF"
+        private const val TAG = "NotificationsFragment"
         fun newInstance() = NotificationsFragment()

         private val notificationDiffCallback: DiffUtil.ItemCallback<NotificationViewData> =
diff --git a/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsViewModel.kt b/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsViewModel.kt
index 3c07ed011..1cf0717d3 100644
--- a/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsViewModel.kt
+++ b/app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsViewModel.kt
@@ -282,6 +282,7 @@ class NotificationsViewModel @Inject constructor(
     private val timelineCases: TimelineCases,
     private val eventHub: EventHub
 ) : ViewModel() {
+    val activeAccount = accountManager.activeAccount

     val uiState: StateFlow<UiState>

@@ -334,9 +335,9 @@ class NotificationsViewModel @Inject constructor(
         viewModelScope.launch {
             uiAction
                 .filterIsInstance<InfallibleUiAction.SaveVisibleId>()
-                .distinctUntilChanged()
+//                .distinctUntilChanged()
                 .collectLatest { action ->
-                    Log.d(TAG, "Saving visible ID: ${action.visibleId}")
+                    Log.d(TAG, "Saving visible ID: ${action.visibleId} to ${activeAccount?.id}")
                     accountManager.activeAccount?.let { account ->
                         account.lastNotificationId = action.visibleId
                         accountManager.saveAccount(account)
```

This adds additional logs to:

- `NotificationsFragment`, log the active account ID when it starts and pauses
- `NotificationsViewModel`, log the active account ID when the change to `lastNotificationId` is saved

(commenting out the `distinctUntilChanged` just means the position will be saved even if it hasn't changed, makes it easier to trigger the problem).

Launch, and sign in with two accounts, and do the following:

1. From account 1, switch to the notifications tab, let notifications load
2. Open the left-hand menu, switch to account 2 (do not leave the notifications tab)

Logs (with line numbers):

```
1 2023-05-29 16:16:28.021 NotificationsFragment             D  onCreateView, active account = 1
[...]
2 2023-05-29 16:17:41.084 AccountManager                    D  setActiveAccount: saving account with id 1
3 2023-05-29 16:17:41.084 AccountManager                    D  saveAccount: saving account with id 1
4 2023-05-29 16:17:41.117 NotificationsFragment             D  onPause, active account = 1
5 2023-05-29 16:17:41.118 NotificationsViewModel            D  Saving visible ID: 153196180 to 1
6 2023-05-29 16:17:41.118 AccountManager                    D  saveAccount: saving account with id 2
```

1: `NotificationsFragment` starts for account 1 (correct)
2-3: The account switch is starting. Account 1 is saved by the AccountManager (correct)
4: `NotificationsFragment` is stopping, its `onPause()` runs, which wants to save the visible notification. The active account is still 1 (correct)
5: `NotificationsViewModel` has received the `SaveVisibleId` message, and is saving the notification ID (153196180) to account 1, calling `AccountManager.saveAccount` (correct)
6: `AccountManager.saveAccount()` is now running, but in the intervening time the account switch has completed, and the active account is now 2. The notification ID is being saved to the wrong account. (very bad)

We know from https://github.com/tuskyapp/Tusky/issues/3689#issuecomment-1565456594 that if a Mastodon server gets a notification ID that looks like a GoToSocial ID it will return the earliest page of notifications.

So if, in the scenario above, account 1 is a GoToSocial account and account 2 is a Mastodon account a GoToSocial notification ID will be saved to the database for the Mastodon account, and chaos ensures.

If this is correct this should be straightforward to reproduce.

**Comment by gudenau** [comment]
This might not be fully correct but I just played with it a little bit. For me switching accounts with the notification tab opened didn't do it by itself but the following did:
- Switch to account with slightly slow server
- Refresh notifications
- Quickly switch to other account
- Notifications go back in time

I only got it to trigger while refreshing when paying attention, I'm not sure if this was a coincidence since this is such an intermittent issue. Going forward I'll try to keep track of how I interact with this.

**Comment by nikclayton** [comment]
This comment contains an APK which will install as "Tusky nightly". It's the code from v22 beta 6 plus the change in https://github.com/tuskyapp/Tusky/pull/3697.

(It's a .zip file because GitHub doesn't let you attached .apk files -- just unzip it).

If you're comfortable side-loading Android apps (instructions at https://www.xda-developers.com/how-to-sideload-install-android-app-apk/) please can you try it (it won't interfere with your existing Tusky (blue icon) app) and let me know? Thanks.


[Tusky_22.0 beta 6-318f4b9c_108_318f4b9c_green_debug.zip](https://github.com/tuskyapp/Tusky/files/11594017/Tusky_22.0.beta.6-318f4b9c_108_318f4b9c_green_debug.zip)

**Comment by nehemiagurl** [comment]
Seems to be working fine!
I even connected another account to see that there aren't any weird multi-account interactions and it still seems fine.

**Comment by yuvalne** [comment]
I do want to mention that this issue was still present in 22.0-beta 7, and is present in the final build of 22.0. So it works on Nightly, but not on the main release.

**Comment by nehemiagurl** [comment]
Hi, issue is still present in 23.0-beta 1 on the main release (but as before, not present on the nightly version, specifically 23.0 beta-1-961de796)
